### PR TITLE
Analysis type

### PR DESCRIPTION
**Pull request by fergusdixon**
## Overview

| Q                 | A
| ----------------- | ---
| Bugfix?           | no
| New feature?      | yes
| Breaking Changes? | no
| Fixed issues      | #52 

## The problem
AnalysisType model was missing
...

## How I resolved it
Added AnalysisType model with mocks and tests
...

Notify the following people: @darrynten

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,9 +3,11 @@ composer.phar
...
 /bin/
 /tests/coverage/
+composer.lock
+clover.xml
+.idea/
 *.log
 
 # Commit your application's lock file http://getcomposer.org/doc/01-basic-usage.md#composer-lock-the-lock-file
 # You may choose to ignore a library lock file http://getcomposer.org/doc/02-libraries.md#lock-file
-composer.lock
```
Comment (fergusdixon) [inline .gitignore]: I changed IDEs and it did this, is that okay?

Comment (darrynten) [reply .gitignore]: You should revert this, it's best to not commit that at this stage of development (we put it back later)

File: .gitignore [hunk 2]
```diff
@@ -3,9 +3,11 @@ composer.phar
 /vendor/
 /bin/
 /tests/coverage/
+composer.lock
+clover.xml
+.idea/
```
Comment (darrynten) [inline .gitignore]: Strange that these moves, but cool - and good one on ignoring `.idea`

File: src/Models/AnalysisType.php [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+     * @var array $features
+     */
+    protected $features = [
+        'all' => true,
+        'get' => true,
+        'save' => true,
+        'delete' => true,
+    ];
+
+    public function saveBatch()
```
Comment (darrynten) [inline src/Models/AnalysisType.php]: What's going on with this?

Missing a docblock, and what does it actually POST through?

File: tests/SageOne/Models/AnalysisTypeModelTest.php [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    public function testCanNotNullify()
+    {
+        $this->verifyCanNotNullify(AnalysisType::class, 'id');
+    }
+
+    public function testAttributes()
+    {
+        $this->verifyAttributes(AnalysisType::class, [
+            'id' => [
+            'type' => 'integer',
```
Comment (darrynten) [inline tests/SageOne/Models/AnalysisTypeModelTest.php]: Indentation

File: tests/SageOne/Models/AnalysisTypeModelTest.php [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            $this->assertEquals(3, $model->analysisCategories->results[1]->id);
+            $this->assertEquals(1, $model->analysisCategories->results[1]->analysisTypeId);
+            $this->assertEquals('sample string 8', $model->analysisCategories->results[1]->description);
+            $this->assertTrue($model->active);
+        });
+    }
+
+    public function testSave()
+    {
+        $this->verifySave(AnalysisType::class, function ($model) {
```
Comment (darrynten) [inline tests/SageOne/Models/AnalysisTypeModelTest.php:110]: Is there a reason this callable is empty?

**Review by darrynten — CHANGES_REQUESTED** [review]
Some minor requests and questions

File: .gitignore (hunk 2 above)
Comment (fergusdixon) [reply .gitignore]: So do I need to revert it?

File: tests/SageOne/Models/AnalysisTypeModelTest.php (hunk 5 above)
Comment (fergusdixon) [reply tests/SageOne/Models/AnalysisTypeModelTest.php:110]: The verifySave populates the object, so I didn't see the point of doing the usual


            $model->id = 1;
            $model->analysisTypeId = 2;
            $model->description = 'sample string 3';
        }, function ($savedModel) {
            $this->assertInstanceOf(AnalysisCategory::class, $savedModel);
            $this->assertEquals(1, $savedModel->id);
            $this->assertEquals(2, $savedModel->analysisTypeId);
            $this->assertEquals('sample string 3', $savedModel->description);
        });
`
Is it necessary for the tests?

Comment (darrynten) [reply tests/SageOne/Models/AnalysisTypeModelTest.php:110]: I'm actually not 100% sure, @make-it-git will know more than me about that

Comment (kiproping) [reply tests/SageOne/Models/AnalysisTypeModelTest.php:110]: I also thing it's redundant. It populates the object second time for no reason.

Comment (fergusdixon) [reply tests/SageOne/Models/AnalysisTypeModelTest.php:110]: Unless it's testing variable assignment?

File: .gitignore (hunk 1 above)
Comment (fergusdixon) [reply .gitignore]: So just revert the -composer.lock?

File: tests/SageOne/Models/AnalysisTypeModelTest.php (hunk 5 above)
Comment (darrynten) [reply tests/SageOne/Models/AnalysisTypeModelTest.php:110]: It could possibly be, perhaps we should wait to hear what @make-it-git has to say about it :D

Comment (make-it-git) [reply tests/SageOne/Models/AnalysisTypeModelTest.php:110]: Yes, it's testing variable assignment

File: src/Models/AnalysisType.php [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    ];
+
+    /**
+     * Saves a collection of Analysis Types.
+     *
+     * @return array
+     */
+    public function saveBatch()
+    {
+        return $this->request->request('POST', $this->endpoint, 'SaveBatch');
```
Comment (darrynten) [inline src/Models/AnalysisType.php]: What gets passed to this request?

**Review by darrynten — CHANGES_REQUESTED** [review]
I'm not sure what the saveBatch method is doing

File: src/Models/AnalysisType.php (hunk 6 above)
Comment (fergusdixon) [reply src/Models/AnalysisType.php]: I wasn't 100% sure how the existing framework we've built handles body vs URI parameters, so I found https://accounting.sageone.co.za/api/1.1.2/Help/Api/POST-SupplierAdjustment-Calculate that does a similar function and imitated it.

File: tests/SageOne/Models/AnalysisTypeModelTest.php [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+            $model->id = 1;
+            $model->description = "sample string 2";
+            $model->active = true;
```
Comment (fergusdixon) [inline tests/SageOne/Models/AnalysisTypeModelTest.php:113]: To set variables in the collection I tried to implement a __set method but I couldn't get it to work, is it worth doing something like:
`$user = createUser();
$role = $user->role;
$role->rolename = 'Test';` to get around that?

**Review by darrynten — DISMISSED** [review]

**Comment by darrynten** [comment]
As discussed with @fergusdixon the way that Sage do the batch saving of AnalysisType - the model is singular but has a batch call on it, and we have not yet encountered the need to have a custom collection of these.

The next PR will be some form of Library Exception getting thrown here, with details on the inconsistency, relevant links to PRs and discussions like this one, and asking for help to fix the problem, so that the next dev who may need to do a batch save has a head start on fixing this.

**Review by darrynten — APPROVED** [review]

**Pull request merged**